### PR TITLE
[dsv4] Bound C4 indexer logits peak memory via varlen routing   query-axis chunking

### PR DESCRIPTION
**Pull request by Leoyzen**
## Problem

DSV4 C4 indexer materializes a `[query_rows, max_c4_seq_len]` fp32 logits tensor
that reaches ~9 GiB at 300K+ context, OOMing under `mem_fraction_static=0.8`.
The crash path is `forward_c4_indexer → deep_gemm.fp8_paged_mqa_logits` (paged
path, SM90/SM100 default).

Related: #33246

## Solution

Port the DSA indexer's budget detection + query-axis chunking pattern to DSV4.
When the estimated logits memory exceeds a budget (reusing DSA's
`_MQA_LOGITS_*` constants and `SGLANG_DSA_MQA_LOGITS_FREE_MEM_FRACTION` — no
new env var), route the computation to the **varlen non-paged path**
(`deep_gemm.fp8_fp4_mqa_logits` with per-row `ks/ke`) and chunk on the query
axis. Normal (in-budget) prefill stays on the fast paged path unchanged.

### Why varlen routing, not paged-path chunking

Paged `fp8_paged_mqa_logits` takes a global `schedule_meta` (`deep_gemm_metadata`)
built from the full batch's `c4_seq_lens` via `get_paged_mqa_logits_metadata`.
Query-axis slicing on the paged kernel requires per-chunk metadata rebuild
(extra JIT kernel launches, CUDA-graph incompatibility, unverified correctness).
The varlen kernel uses per-row `ks/ke` cu_seqlen arrays — slicing is trivially
safe, no global schedule to rebuild.

### Why per-request processing for multi-request batches

`topk_transform_512` assumes logits indexed in per-request page space
`[0, c4_seq_len)`. A multi-request varlen buffer concatenates KV, so request B's
logits would be indexed `[len_A, len_A+len_B)` — mismatch with topk mask
`[0, len_B)`. DSA's topk accepts `ks` for offset compensation; DSV4's does not.
Per-request processing avoids this entirely. Single-request oversize (the most
common case: one 300K+ prompt) goes through directly without the loop.

### FP8/FP4 unification

`_forward_nonpaged_indexer` now uses `deep_gemm.fp8_fp4_mqa_logits` for both FP8
(`q_sf=None`) and FP4 (`q_sf=tensor`). The legacy `fp8_mqa_logits` is a C++
wrapper that calls `fp8_fp4_mqa_logits` with `q_sf=nullopt` — same kernel, zero
performance difference. This also enables FP4 models to benefit from the oversize
routing (previously FP4 was rejected by `_can_use_nonpaged_indexer`).

### Precedent

- **vLLM #36178**: Always uses varlen `fp8_fp4_mqa_logits` for prefill (paged for
 decode only), with `VLLM_SPARSE_INDEXER_MAX_LOGITS_MB` budget + chunking.
- **sglang DSA indexer**: Already has `_get_mqa_logits_budget_bytes` +
 `_should_chunk_mqa_logits` + `_get_topk_ragged` (query-axis chunked loop).
 DSV4 indexer was missing this pattern.

## Changes

| File | Change |
|------|--------|
| `python/sglang/srt/layers/attention/dsv4/indexer.py` | +441/-52: budget helpers, varlen routing, chunk loop, FP8/FP4 unification, CUDA-graph guards |
| `test/registered/unit/layers/test_dsv4_nonpaged_indexer.py` | +332/-52: 4 test classes, 13 test methods |

## Known limitations

- **Context parallel**: CP + oversize stays on paged path (cross-rank KV gather
 out of scope)
- **Non-deep-gemm backends** (torch fallback, tilelang, aiter): Stay on original
 path (no varlen equivalent). SM120 torch fallback is addressed by #33259.
- **CUDA-graph capture**: Stays on paged single-pass (shapes are pre-validated)

## Relationship to #33259

Complementary, not overlapping:
- **This PR** fixes the deep_gemm path (SM90/SM100 — the OOM source in production)
- **#33259** fixes the SM120 torch fallback path (seq-axis chunking)

## Test plan

- [x] `pre-commit run` (isort, ruff, black, codespell) — all passed
- [ ] `base-a-test-cpu` CI (mock-based unit tests)
- [ ] Manual test: 300K+ context prefill on L20X ×4 with `mem_fraction_static=0.8`
      previously OOM at ~280K)
- [ ] Regression: normal prefill (short context) bit-identical to pre-change behavior

---
### CI States

Latest PR Test (Base): :x: [Run #31119319356](https://github.com/sgl-project/sglang/actions/runs/31119319356)
Latest PR Test (Extra): :x: [Run #31119319531](https://github.com/sgl-project/sglang/actions/runs/31119319531)

### CODE REVIEW

File: python/sglang/srt/layers/attention/dsv4/indexer.py [hunk 1]
```diff
@@ -630,6 +712,228 @@ def _forward_nonpaged_indexer(
...
+                c4_seq_lens[req_start:req_end].reshape(-1).to(torch.int32).contiguous()
+            )
+            gather_seq_lens = req_ke[-1:]
+            req_ks = torch.zeros_like(req_ke)
+            max_seqlen_k = (
+                (final_c4_len + c4_page_size - 1) // c4_page_size * c4_page_size
+            )
+
+            # Gather contiguous KV for this request.
+            k_u8, scale_u8 = token_to_kv_pool.get_index_k_scale_buffer(
```
Comment (hnyls2002) [inline python/sglang/srt/layers/attention/dsv4/indexer.py:816]: This is `_forward_nonpaged_indexer`'s gather (L684-690) with every `plan.*` swapped for a loose local, so it bypasses `_get_nonpaged_indexer_plan` and with it `_can_use_nonpaged_indexer` -- hisparse, non-EXTEND modes, TBO, graph capture and the master env flag are all unguarded here.

L807-815 rebuilds all 8 `NonPagedIndexerPlan` fields verbatim, just rebased to `req_start`. Make that function return a per-request list instead of opening a second path.

File: python/sglang/srt/layers/attention/dsv4/indexer.py [hunk 2]
```diff
@@ -500,8 +571,7 @@ def _can_use_nonpaged_indexer(
         ):
             return False
         if (
-            c4_indexer.use_fp4_indexer
-            or envs.SGLANG_OPT_USE_TILELANG_INDEXER.get()
+            envs.SGLANG_OPT_USE_TILELANG_INDEXER.get()
```
Comment (hnyls2002) [inline python/sglang/srt/layers/attention/dsv4/indexer.py]: Dropping `use_fp4_indexer` routes FP4 into `get_index_k_scale_buffer`, but that gather is fp8-only: `_get_k_and_s_triton` hardcodes `index_head_dim + 4` bytes/token while the FP4 pool packs `index_head_dim // 2 + 4`. Reading a 68B buffer at a 132B stride is silently wrong, no crash. Keep FP4 rejected.

File: python/sglang/srt/layers/attention/dsv4/indexer.py [hunk 3]
```diff
@@ -630,6 +712,228 @@ def _forward_nonpaged_indexer(
...
+        # are allocated between the query and the logits allocation.
+        if get_is_capture_mode():
+            # During CUDA-graph capture we cannot call mem_get_info; fall
+            # back to the (already conservative) static budget.
+            realtime_budget = budget_bytes
+        else:
+            free_mem, _ = torch.cuda.mem_get_info(device.index)
+            total_mem = torch.cuda.get_device_properties(device.index).total_memory
+            static_cap = int(total_mem * _MQA_LOGITS_TOTAL_MEM_FRACTION)
+            realtime_budget = min(int(free_mem * 0.5), static_cap)
```
Comment (hnyls2002) [inline python/sglang/srt/layers/attention/dsv4/indexer.py]: `budget_bytes` is dead now -- its only reader is behind `get_is_capture_mode()`, which the caller already guarantees is False. That also disables the tests: `_run_chunked_path` varies only `budget_bytes`, so the chunked-vs-unchunked test compares two identical unchunked runs.

File: python/sglang/srt/layers/attention/dsv4/indexer.py (hunk 1 above)
Comment (Leoyzen) [reply python/sglang/srt/layers/attention/dsv4/indexer.py:816]: Yeah fair — the oversize path wasn't checking any of that. I've added the env flag plus the TBO/hisparse/piecewise/breakable guards to the routing block, so it now bails in exactly the cases `_can_use_nonpaged_indexer` would. Also cleaned up a leftover `if use_fp4_indexer` branch that was unreachable after that.

On the per-request plan refactor: I think that's the right shape long-term, but `_get_nonpaged_indexer_plan` is built around `batch_size == 1` and loosening it is a bigger change — would rather do it as a follow-up so this fix stays reviewable.

File: python/sglang/srt/layers/attention/dsv4/indexer.py (hunk 2 above)
Comment (Leoyzen) [reply python/sglang/srt/layers/attention/dsv4/indexer.py]: Good catch, that one's on me — it was silent corruption, not a fail-fast. The FP8-stride gather walks past the 68B/token FP4 rows and reads garbage without complaining.

Restored `use_fp4_indexer` in the rejection list, and also kept FP4 out of the oversize chunk path for the same reason, so it stays on the paged path where the layout math is already correct. Making the varlen gather FP4-aware is tracked as a follow-up.

File: python/sglang/srt/layers/attention/dsv4/indexer.py (hunk 3 above)
Comment (Leoyzen) [reply python/sglang/srt/layers/attention/dsv4/indexer.py]: Both fair. `budget_bytes` was indeed only read under capture mode, which never fires there — so it did nothing. It's now the actual chunking limit, clamped by real free memory on CUDA so we keep the OOM safety.

And the chunked-vs-unchunked test now genuinely chunks (one pass vs 3-rows-per-chunk) — before, both runs were the same code path, so it "passed" for no reason. It also used to crash on CPU via `mem_get_info` with no CUDA; that's guarded now too.
